src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/DocumentMapping/RemoteDocumentMappingService.cs

```diff
@@ -5,22 +5,82 @@
 using System.Composition;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor;
 using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.CodeAnalysis.Razor;
 using Microsoft.CodeAnalysis.Razor.DocumentMapping;
 using Microsoft.CodeAnalysis.Razor.Logging;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
 using Microsoft.CodeAnalysis.Razor.Workspaces;
+using Microsoft.CodeAnalysis.Remote.Razor.ProjectSystem;
+using Microsoft.CodeAnalysis.Text;
 
 namespace Microsoft.CodeAnalysis.Remote.Razor.DocumentMapping;
 
 [Export(typeof(IDocumentMappingService)), Shared]
 [method: ImportingConstructor]
 internal sealed class RemoteDocumentMappingService(
     IFilePathService filePathService,
+    IProjectCollectionResolver projectCollectionResolver,
     ILoggerFactory loggerFactory)
     : AbstractDocumentMappingService(filePathService, loggerFactory.GetOrCreateLogger<RemoteDocumentMappingService>())
 {
-    protected override ValueTask<RazorCodeDocument?> TryGetCodeDocumentAsync(Uri razorDocumentUri, CancellationToken cancellationToken)
+    private readonly IProjectCollectionResolver _projectCollectionResolver = projectCollectionResolver;
+
+    public async Task<(Uri MappedDocumentUri, LinePositionSpan MappedRange)> MapToHostDocumentUriAndRangeAsync(
+        RemoteDocumentSnapshot originSnapshot,
+        Uri generatedDocumentUri,
+        LinePositionSpan generatedDocumentRange,
+        CancellationToken cancellationToken)
     {
-        throw new NotImplementedException();
+        var razorDocumentUri = FilePathService.GetRazorDocumentUri(generatedDocumentUri);
```

Comment:
Nothing to do now, but just a note that this feels fragile, and might not be possible with the source generator hooked up. I suspect `FilePathService` will be useless, if not actively harmful, in cohosting. Having said that, it probably depends on where `generatedDocumentUri` comes from, because until we have the source generator hooked up the old conventions still apply, but they might end up being false positives.

Reply:
I agree. For co-hosting, I don't think we'll want `IFilePathService` at all. Instead, we should probably be able to determine the original razor document from the `SourceGeneratedDocument`.